internal/runners/push/push.go

```diff
@@ -170,61 +174,93 @@ func (r *Push) Run(params PushParams) (rerr error) {
-
-	// Detect the target branch
-	var branch *mono_models.Branch
-	if projectCreated || r.project.BranchName() == "" {
-		// https://www.pivotaltracker.com/story/show/176806415
-		// If we have created an empty project the only existing branch will be the default one
+		// Fetch the newly created project's default branch (for updating activestate.yaml with).
+		targetPjm, err = model.LegacyFetchProjectByName(targetNamespace.Owner, targetNamespace.Project)
+		if err != nil {
+			return errs.Wrap(err, "Failed to fetch newly created project")
+		}
 		branch, err = model.DefaultBranchForProject(targetPjm)
 		if err != nil {
 			return errs.Wrap(err, "Project has no default branch")
 		}
+
+		projectCreated = true
+
 	} else {
-		branch, err = model.BranchForProjectByName(targetPjm, r.project.BranchName())
-		if err != nil {
-			return errs.Wrap(err, "Could not get branch %s", r.project.BranchName())
-		}
-	}
 
-	// Check if branch is already up to date
-	if branch.CommitID != nil && branch.CommitID.String() == commitID.String() {
-		return errNoChanges
-	}
+		// Now we get to the actual push logic
+		r.out.Notice(locale.Tl("push_to_project", "Pushing to project [NOTICE]{{.V1}}[/RESET] under [NOTICE]{{.V0}}[/RESET].", targetNamespace.Owner, targetNamespace.Project))
 
-	// Check whether there is a conflict
-	if branch.CommitID != nil {
-		mergeStrategy, err := model.MergeCommit(*branch.CommitID, commitID)
-		if err != nil {
-			if errors.Is(err, model.ErrMergeCommitInHistory) {
-				return errNoChanges
+		// Detect the target branch
+		if r.project.BranchName() == "" {
+			branch, err = model.DefaultBranchForProject(targetPjm)
+			if err != nil {
+				return errs.Wrap(err, "Project has no default branch")
+			}
+		} else {
+			branch, err = model.BranchForProjectByName(targetPjm, r.project.BranchName())
+			if err != nil {
+				return errs.Wrap(err, "Could not get branch %s", r.project.BranchName())
 			}
-			if !errors.Is(err, model.ErrMergeFastForward) {
-				if params.Namespace.IsValid() {
-					return errTargetInvalidHistory
+		}
+
+		// Check if branch is already up to date
+		if branch.CommitID != nil && branch.CommitID.String() == commitID.String() {
+			return errNoChanges
+		}
+
+		// Check whether there is a conflict
+		if branch.CommitID != nil {
+			mergeStrategy, err := model.MergeCommit(*branch.CommitID, commitID)
+			if err != nil {
+				if errors.Is(err, model.ErrMergeCommitInHistory) {
+					return errNoChanges
 				}
-				return errs.Wrap(err, "Could not detect if merge is necessary")
+				if !errors.Is(err, model.ErrMergeFastForward) {
+					if params.Namespace.IsValid() {
+						return errTargetInvalidHistory
+					}
+					return errs.Wrap(err, "Could not detect if merge is necessary")
+				}
+			}
+			if mergeStrategy != nil {
+				return errPullNeeded
 			}
 		}
-		if mergeStrategy != nil {
-			return errPullNeeded
+
+		// Perform the push.
+		newCommitID, err = bp.AttachStagedCommit(&model.AttachStagedCommitParams{
```

Comment:
Reading this makes me think that `AttachStagedCommit` is returning a new commit ID when it should return the commit ID of the staged commit, which we already have. Is that not the case? If it is, what is the value of having `AttachStagedCommit` return a commit ID at all? I think an error might be enough.

Reply:
Good point. I've verified that `newCommitID` is the same as commitID, so I've just made `AttachStagedCommit()` return an error only.